pkg/middleware/region_proxy.go

```diff
@@ -6,33 +6,65 @@ import (
 	"net/http"
 	"net/http/httputil"
 	"net/url"
+	"strings"
+	"sync"
 
 	sdk "github.com/openshift-online/ocm-sdk-go"
 	"github.com/openshift-online/ocm-sdk-go/logging"
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+const (
+	// region ids
+	APSoutEast1 = "ap-southeast-1"
+	USEast1     = "us-east-1"
+	Global      = "global"
+)
+
 var requestsDispatched = prometheus.NewCounter(
 	prometheus.CounterOpts{
 		Name: "region_proxy_dispatch_total",
 		Help: "The total number of requests dispatched by region proxy.",
 	},
 )
 
-type getDispatchHostFunc func(context.Context, logging.Logger, *http.Request, *sdk.Connection) (string, error)
+type getClusterIdsHandlerFunc func(context.Context, logging.Logger, *http.Request) (ClusterIds, error)
+
+type checkLocalHandlerFunc func(context.Context, logging.Logger, ClusterIds) (bool, error)
 
-type errorHandler func(http.ResponseWriter, *http.Request, error)
+type dispatchHandlerFunc func(context.Context, logging.Logger, http.ResponseWriter, *http.Request,
+	http.Handler, string, map[string]sdk.Region) error
+
+type errorHandlerFunc func(http.ResponseWriter, *http.Request, error)
+
+type ClusterIds struct {
+	Id         string
+	ExternalId string
+}
+
+type ClusterCache struct {
```

Comment:
I'd be careful with this. Over time this cache could grow quite large. You should consider some sort if LRU cache with a max size (don't build this yourself) to prevent memory leaks. CS in particular is quite sensitive about memory right now, this will make them worried about the proxy potentially impacting the overall service.

Reply:
@macgregor 
I changed to use LRU cache
please help to review the latest commit, thanks